ZEP008-hot-backup.md

```diff
@@ -0,0 +1,619 @@
+
+Scenario 3: (this case is very less likely)
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- BackupActor starts taking checkpoint
+- BackupActor completes taking the checkpoint
+- Leader change before "checkpoint taken" is committed
+
+This scenario is very rare because usually the committing `checkpoint taken` record will be much faster than taking a backup.
+In this case, the new leader process the checkpoint command and triggers the backup again. But a completed backup Backup[N][p][old leader] exists. The new leader will start taking the backup to a new location Backup[N][p][new leader] and will commit the new one. This is a bit tricky. If the new leader commits a backup with the same id, it is semantically equivalent to the backup from the old leader. So there is no harm in allowing two backups for the same partition. However, this could be confusing and consumes extra storage which is unnecessary. But since it doesn't lead to any inconsistent state, and this scenario is very rare, we can allow this to happen to keep it simple.
+
+Scenario 4:
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- BackupActor starts taking snapshot
+- checkpoint taken record is committed
+- Failover happens before backup is completed
+
+In this scenario, the new leader will not process the checkpoint command. It will only replay `checkpoint taken` record. When replaying it updates the checkpointId and position, but will not trigger a backup. As a result, we have a partial backup for this partition initiated by the old leader. Since the new leader does not take a backup, the backup will never be completed. The coordinator or a user monitoring the backup will observe it as "ongoing" for ever. Inorder to prevent that, we have added a cleanup after the failover. After failover, once the replay is completed, the `BackupActor` marks the previous checkpoint as failed if it is still marked as ongoing.
+
+Note:- In some cases, it might be possible for the new leader to take a consistent backup. But to keep it simple we fail the backup.
+
+Scenario 5:
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- BackupActor starts taking snapshot
+- checkpoint taken record is committed
+- New snapshot is taken with processedposition > checkpoint position
+- Failover happens before backup is completed
+- New leader starts with the new snapshot at processedPosition > checkpoint position
+
+This is similar to scenario 4. The difference is that the checkpoint taken is not replayed. But everything else is similar. In this case, there is no way for the new leader to take a consistent backup because the snapshot is already past the checkpoint record and it is probably already compacted.
```

Comment:
:thought_balloon: If the checkpoint ID is part of the snapshot, then theoretically we could perform the same clean up? If we know the last checkpoint ID part of snapshot X is Y, then we can also clean up previous ongoing backups, no?

Reply:
By cleanup do you mean deleting old backups of marking old ongoing backups as failed?